### PR TITLE
1202 select testimonials

### LINKED ISSUE #1202: 🐞 Events - Unable to select specific testimonials
When I setup an event, I want to select the correct testimonials e.g. when I edit the CA Workshop event, I don't want to see Firebootcamp testimonials

```[tasklist]
### Tasks
- [x] Add ability to select specific testimonials (like the consulting pages)
```

Image

### PR DESCRIPTION
**Pull request by johnxu16**
- Add Selection of Testimonial Categories
- Update schema

Fixes #1202

Affected routes:

- /admin
- /consulting/*
- /events/*
- /training/*

Add done video, screenshots

image
**Figure: Added option for adding testimonials categories**

### CODE REVIEW

File: helpers/getTestimonials.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    (t) => t.node
+  );
+
+  testimonialsResult = testimonialsResult.sort(() => 0.5 - Math.random());
+
+  // Adds general testimonials if not filled by testimonials with matching categories
+  if (testimonialsResult.length < 3) {
+    const generalTestimonials = await client.queries.testimonalsQuery({
+      categories: defaultCategory ?? "General",
+    });
```
Comment (bradystroud) [inline helpers/getTestimonials.ts]: I dont think we need 3 testimonials, 1 or 2 looks ok
(similar to presenters)

image

**Review by bradystroud — CHANGES_REQUESTED** [review]
Looks good John, can you film a done video on how this works?

File: helpers/getTestimonials.ts (hunk 1 above)
Comment (johnxu16) [reply helpers/getTestimonials.ts]: Yes! Forgot we had discussion about this during migrating events page

Comment (johnxu16) [reply helpers/getTestimonials.ts]: Done - use default testimonials only when there is not testimonial found with graphql

**Comment by bradystroud** [comment]
@johnxu77 I cannot get this working

image
**Figure: Web dev selected, but I'm seeing the Internship testimonials**

File: .tina/collections/events.tsx [hunk 2]
```diff
@@ -41,6 +41,25 @@ export const eventsSchema: Collection = {
       name: "showTestimonials",
       label: "Show Testimonials",
     },
+    {
+      type: "object",
+      label: "Testimonial Categories",
+      name: "testimonialCategories",
+      ui: {
+        itemProps(item) {
+          return { label: item.testimonialCategory }
```
Comment (Harry-Ross) [inline .tina/collections/events.tsx]: Item can be null here, can you add a fallback label?

File: content/events/angular-workshop.mdx [hunk 3]
```diff
@@ -109,8 +109,13 @@ _body:
             icon="/images/icons/university_hat.png"
           />
     _template: Agenda
+showTestimonials: false
+testimonialCategories:
+  - testimonialCategory: content/testimonialCategories/Chinafy.mdx
```
Comment (Harry-Ross) [inline content/events/angular-workshop.mdx]: Remove this testimonial category

File: helpers/getTestimonials.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import client from "../.tina/__generated__/client";
+
+export const GetTestimonialsByCategories = async (
+  categories,
+  defaultCategory
```
Comment (Harry-Ross) [inline helpers/getTestimonials.ts]:
```suggestion
  categories: string[],
  defaultCategory?: string

```

**Review by Harry-Ross — CHANGES_REQUESTED** [review]
Looks good to me, apart from my small changes! Very happy that you have moved that messy testimonial fetching stuff to its own file

File: helpers/getTestimonials.ts (hunk 4 above)
Comment (johnxu16) [reply helpers/getTestimonials.ts]: Done - type added and changed defaultCategory a optional param

File: content/events/angular-workshop.mdx (hunk 3 above)
Comment (johnxu16) [reply content/events/angular-workshop.mdx]: Done - removed

File: .tina/collections/events.tsx (hunk 2 above)
Comment (johnxu16) [reply .tina/collections/events.tsx]: Done - added a fallback label as "Select your testimonial category"

**Comment by johnxu16** [comment]
> @johnxu77 I cannot get this working
> 
> image **Figure: Web dev selected, but I'm seeing the Internship testimonials**

It got me once before. There is no testimonial with the category 'Web development'. As a fallback, it shows testimonials in the 'Internship' category.


**Figure: No testimonials with category 'Web development'**


**Figure: fallback code**

**Comment by wicksipedia** [comment]
if there are no testimonials in the category - then none should be shown

**Comment by Harry-Ross** [comment]
> if there are no testimonials in the category - then none should be shown

I think that it should always fall back to general (like the original logic), and remove the second argument from the function

**Comment by Harry-Ross** [comment]
image

Is this still the expected behaviour? I think it should be showing internship testimonials if I select it.

**Comment by johnxu16** [comment]
> image
> 
> Is this still the expected behaviour? I think it should be showing internship testimonials if I select it.

I double-checked the website on production, it worked as before.

I am thinking of how to solve it.

**Review by bradystroud — CHANGES_REQUESTED** [review]
This is looking good 🙂 

When there are only 2 testimonials, they should be in the centre, not left aligned

image

**Comment by johnxu16** [comment]
> This is looking good 🙂
> 
> When there are only 2 testimonials, they should be in the centre, not left aligned
> 
> image

Done - component updated
Plus, I refactored the 'useEdit' in the component 'TestimonialRow', so it works on other pages as well.

**Review by bradystroud — APPROVED** [review]
LGTM

**Review by Harry-Ross — APPROVED** [review]
LGTM 🛥️

**Pull request merged**